File: src/i1tbfssymmetric/CS/TBFSSymmetric.Algo/CustomQueue.cs
namespace TBFSSymmetric.Algo;

public class CustomQueue<T>
{
    private QueueNode<T>? _head;
    private QueueNode<T>? _tail;
    private int _size;

    public int Size => _size;

    public bool IsEmpty => _size == 0;

    public T Peek()
    {
        if (IsEmpty)
        {
            throw new InvalidOperationException("Queue is empty.");
        }

        return _head!.Value;
    }

    public void Enqueue(T value)
    {
        QueueNode<T>? temp = new(value);
        if (_head == null)
        {
            _head = temp;
            _tail = temp;
        }
        else
        {
            _tail!.Next = temp;
            _tail = temp;
        }
        _size++;
    }

    public T Dequeue()
    {
        if (IsEmpty)
        {
            throw new InvalidOperationException("Queue is empty.");
        }
        T? value = _head!.Value;
        _head = _head.Next;
        _size--;
        return value;
    }

    public override string ToString()
    {
        if (Size == 0) return "[]";
        QueueNode<T>? temp = _head;
        string? outStr = "[";
        while (temp != null)
        {
            outStr += $"{temp.Value}, ";
            temp = temp.Next;
        }
        return outStr.TrimEnd(',', ' ') + "]";
    }
}

File: src/i1tbfssymmetric/Java/com/example/tbfssymmetric/CustomQueue.java
package com.example.tbfssymmetric;

import java.util.LinkedList;
import java.util.Queue;

public class CustomQueue<T> {
    private Queue<T> queue;

    public CustomQueue() {
        this.queue = new LinkedList<>();
    }

    public int size() {
        return queue.size();
    }

    public boolean isEmpty() {
        return queue.isEmpty();
    }

    public void enqueue(T value) {
        queue.add(value);
    }

    public T dequeue() {
        return queue.poll();
    }

    public T peek() {
        return queue.peek();
    }

    @Override
    public String toString() {
        return queue.toString();
    }
}
